src/Client/Updater.php

```diff
@@ -417,19 +417,24 @@ private function getUpdateStartTime(): \DateTimeImmutable
      * @param bool $terminated
      *   (optional) For internal recursive calls only. This will be set to true if a terminating delegation is found in
      *   the search.
+     * @param array $allDelegatedRoles
+     *   (optional) For internal recursive calls only to avoid looking up the delegated roles every iteration.
      *
      *
      * @return \Tuf\Metadata\TargetsMetadata|null
      *   The target metadata that contains the metadata for the target or null if the target is not found.
      */
-    private function searchDelegatedRolesForTarget(TargetsMetadata $targetsMetadata, string $target, array $searchedRoles, bool &$terminated = false): ?TargetsMetadata
+    private function searchDelegatedRolesForTarget(TargetsMetadata $targetsMetadata, string $target, array $searchedRoles, bool &$terminated = false, array $allDelegatedRoles = []): ?TargetsMetadata
     {
         foreach ($targetsMetadata->getDelegatedKeys() as $keyId => $delegatedKey) {
             $this->signatureVerifier->addKey($keyId, $delegatedKey);
         }
+        if (!\count($allDelegatedRoles)) {
+            $allDelegatedRoles = $targetsMetadata->getDelegatedRoles();
+        }
```

Comment:
It has been a while since I looked at this but is this safe to do?. If we call `getDelegatedRoles()` once for the first `$targetsMetadata` but then in recursive calls when have a different `$delegatedTargetsMetadata` use the same `$allDelegatedRoles` won't this not work for any nested delegations defined in the delegated targets metadata itself.

Reply:
Pushed a new commit that adds a class property to save recalculation instead of passing things recursively.